Allow hiding extra env variables in clean_env()

baldassarreFe opened the pull request:
**Context:** my training job, running in SLURM, periodically submits evaluation jobs, also to be run in SLURM. Both training and evaluation jobs make use of `torch.compile()` which in turn enables torchinductor and triton.

**Problem:** at the first use, torchinductor and triton set the `TRITON_CACHE_DIR` and `TORCHINDUCTOR_CACHE_DIR` environment variables to indicate the location of their cache dirs. When an evaluation job is submitted, it inherits these variables, it tries to access those folders and gets a permission error:
```
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
PermissionError: [Errno 13] Permission denied: '/scratch/slurm_tmpdir/4187753'
...
srun: error: hostname1058: tasks 0,2-7: Exited with exit code 1
srun: Terminating StepId=4197552.0
slurmstepd: error: *** STEP 4197552.0 ON hostname1058 CANCELLED AT 2024-07-25T22:08:52 ***
srun: error: hostname1058: task 1: Exited with exit code 1
```

Note that the job id of the evaluation job is `4197552` but it tries to access `/scratch/slurm_tmpdir/4187753` that is the cache dir derived from the parent training job `4187753`.

**Proposed solution:** submitit already exposes the `clean_env()` context manager to allow hiding some variables when submitting jobs programmatically from within another job. It's easy to add a parameter to `clean_env()` to allow users to hide additional variables that are relevant for their case.

Variable names and docstring are open to suggestions.

jrapin commented on submitit/helpers.py:
```diff
@@ -293,10 +293,16 @@ def monitor_jobs(
 
 @contextlib.contextmanager
-def clean_env() -> tp.Iterator[None]:
+def clean_env(extra_names: tp.Optional[tp.Sequence[str]] = None) -> tp.Iterator[None]:
```
as a tuple is a sequence and is not mutable, you can initialize with an empty tuple and avoid dealing with the None case
```suggestion
def clean_env(extra_names: tp.Optional[tp.Sequence[str]] = ()) -> tp.Iterator[None]:
```

baldassarreFe replied:
yes, it makes sense, just force of habit

jrapin commented on submitit/helpers.py:
```diff
@@ -308,11 +314,17 @@ def clean_env() -> tp.Iterator[None]:
     with submitit.helpers.clean_env():
         executor.submit(...)
     """
+    if extra_names is None:
+        extra_names = ()
```
```suggestion
```

jrapin commented on submitit/helpers.py:
```diff
@@ -308,11 +314,17 @@ def clean_env() -> tp.Iterator[None]:
     """
+    if extra_names is None:
+        extra_names = ()
     distrib_names = ("MASTER_ADDR", "MASTER_PORT", "RANK", "WORLD_SIZE", "LOCAL_RANK", "LOCAL_WORLD_SIZE")
```
you could just set `forbidden_names = ("MASTER_ADDR",...) + tuple(extra_names)` then use it, that seems simpler than the 2 `or`. Then again, it's about the same so whichever seems best to you

baldassarreFe replied:
There is no performance concern here, so I think keeping the two lists separate is better for clarity

jrapin reviewed (approved):
please address comment on the default arg before merging (the other one depends on what you prefer)

baldassarreFe commented:
> please address comment on the default arg before merging (the other one depends on what you prefer)

Default args: I made it an empty tuple, removed the None check, updated type annotation and docstring

Merging the tuples: I think it's clearer to keep them separate, easier to guess what the code does at a glance
